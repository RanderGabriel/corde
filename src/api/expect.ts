import { expectMatchWithNot, MatchesWithNot } from "./matcher";

/**
 * Receives wich command will be tested.
 *
 * @param commandName Command name.
 *
 * @description Do not inform the command prefix if
 * it's already informed in **configs**
 *
 * @returns The **Compare** object, where will handle
 * the type of response is expected.
 */
export function expect(commandName: string): MatchesWithNot {
  return expectMatchWithNot(commandName);
}
