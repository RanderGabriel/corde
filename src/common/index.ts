export * from './cordeBot';
export * from './runtime';
export * from './testColletor';
