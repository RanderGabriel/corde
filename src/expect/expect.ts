import { VoidLikeFunction } from "../types";
import { ExpectMatchesWithNot, MatchWithNot } from "./matcher";

/**
 * Invalid declaration of a test matcher.
 *
 * @since 3.0
 */
export function expect(): void;
/**
 * Receives which command will be tested.
 *
 * Do not inform the command prefix if
 * it's already informed in **configs**
 *
 * @param commandNameResolvable Command name. (Empty strings will resolve failed test)
 *
 * @returns An object with all possible tests to be done
 * in the bot.
 *
 * @since 1.0
 */
export function expect<T extends VoidLikeFunction | number | string>(
  commandNameResolvable: T,
): MatchWithNot;
export function expect<T extends VoidLikeFunction | number | string>(
  commandNameResolvable?: T,
): MatchWithNot | void {
  if (commandNameResolvable == null) {
    return;
  }
  return new ExpectMatchesWithNot(commandNameResolvable);
}
