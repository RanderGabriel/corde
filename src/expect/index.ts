import { RoleMatches, MessageMatches } from "./matcher";
import { Expect, MacherContructorArgs } from "./types";

function getMessageMatchers(): string[] {
  return getFunctions(MessageMatches);
}

function getRoleMatchers(): string[] {
  return getFunctions(RoleMatches);
}

function getFunctions<T>(type: new (...args: any[]) => T): string[] {
  return Object.getOwnPropertyNames(type.prototype).filter(
    (propName) => propName !== "constructor" && typeof type.prototype[propName] === "function",
  );
}

const expectation = {
  not: {},
  inChannel: {},
  inGuild: {},
};

function createTestFunction(classType: any, params: MacherContructorArgs, functionName: string) {
  return (...args: any[]) => (new classType(params) as any)[functionName](...args);
}

function createTestsFromMatches<T>(
  names: string[],
  params: MacherContructorArgs,
  classType: new (...args: any[]) => T,
): [string, (...args: any[]) => any][] {
  return names.map((name) => {
    return [name, createTestFunction(classType, params, name)];
  });
}

function set(from: [string, (...args: any[]) => any][], to: any) {
  from.forEach((test) => {
    to[test[0]] = test[1];
  });
}

const messageTestNames = getMessageMatchers();
const roleMatchers = getRoleMatchers();

const _expect: any = <T extends (() => number | string) | number | string>(commandName: T) => {
  const messageTests = createTestsFromMatches(
    messageTestNames,
    { commandName, isNot: false },
    MessageMatches,
  );

  const isNotMessageTests = createTestsFromMatches(
    messageTestNames,
    { commandName, isNot: true },
    MessageMatches,
  );

  const roleTests = createTestsFromMatches(
    roleMatchers,
    { commandName, isNot: false },
    RoleMatches,
  );

  const isNotRoleTests = createTestsFromMatches(
    roleMatchers,
    { commandName, isNot: true },
    RoleMatches,
  );

  set(messageTests, expectation);
  set(isNotMessageTests, expectation.not);

  set(roleTests, expectation);
  set(isNotRoleTests, expectation.not);

  expectation.inChannel = (channelId: string) => {
    const channelMatchers: any = {
      not: {},
    };

    const messageTests = createTestsFromMatches(
      messageTestNames,
      { commandName, channelId },
      MessageMatches,
    );

    const isNotmessageTests = createTestsFromMatches(
      messageTestNames,
      { commandName, channelId, isNot: true },
      MessageMatches,
    );

    set(messageTests, channelMatchers);
    set(isNotmessageTests, channelMatchers.not);

    return channelMatchers;
  };

  expectation.inGuild = (guildId: string) => {
    const guildMatchers: any = {
      not: {},
    };

    const roleTests = createTestsFromMatches(
      messageTestNames,
      { commandName, guildId },
      RoleMatches,
    );

    const isNotRoleTests = createTestsFromMatches(
      messageTestNames,
      { commandName, guildId },
      RoleMatches,
    );

    set(roleTests, guildMatchers);
    set(isNotRoleTests, guildMatchers.not);

    return guildMatchers;
  };

  return expectation;
};

const messageTests = createTestsFromMatches(
  messageTestNames,
  { commandName: "", isNot: false, isCascade: true },
  MessageMatches,
);

const isNotMessageTests = createTestsFromMatches(
  messageTestNames,
  { commandName: "", isNot: true, isCascade: true },
  MessageMatches,
);

const roleTests = createTestsFromMatches(
  roleMatchers,
  { commandName: "", isNot: false, isCascade: true },
  RoleMatches,
);

const isNotRoleTests = createTestsFromMatches(
  roleMatchers,
  { commandName: "", isNot: true, isCascade: true },
  RoleMatches,
);

_expect.not = {};

set(messageTests, _expect);
set(isNotMessageTests, _expect.not);

set(roleTests, _expect);
set(isNotRoleTests, _expect.not);

const expect = _expect as Expect;

export { expect };
