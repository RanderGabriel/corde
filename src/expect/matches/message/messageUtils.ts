import assert from "assert";
import { Message, MessageEmbed } from "discord.js";
import { messageType, MinifiedEmbedMessage } from "../../../types";
import { isPrimitiveValue, pick } from "../../../utils";

class MessageUtilsManager {
  public messagesMatches(
    returnedMessage: Message,
    expectation: string | number | boolean | MessageEmbed,
  ) {
    let msg = "";
    if (isPrimitiveValue(expectation)) {
      const formattedMsg = this.getMessageByType(returnedMessage, "text") as Message;
      msg = formattedMsg.content;
      return msg === expectation;
    }

    const jsonMessage = this.getMessageByType(returnedMessage, "embed") as MinifiedEmbedMessage;
    msg = JSON.stringify(jsonMessage);
    let result = true;
    try {
      assert.deepStrictEqual(expectation.toJSON(), jsonMessage);
    } catch (error) {
      result = false;
    }
    return result;
  }

  /**
   * Format Discord responses
   *
   * @param answer Discord response for a message sent
   *
   * @param type Type expected of that message
   *
   * @description Discord adds some attributes that are not present in embed message before it is sent
   *
   *  This is data **before** send to Discord
   *
   *  ```javascript
   *   "image": {
   *       "url": "https://i.imgur.com/wSTFkRM.png"
   *   },
   *   "thumbnail": {
   *       "url": "https://i.imgur.com/wSTFkRM.png"
   *   }
   *  ```
   *
   *  And this is part of embed message **after** get from Discord
   *
   *  ```javascript
   *   "image": {
   *     "height": 0,
   *     "proxyURL": "https://images-ext-2.discordapp.net/external/DoAGN014Q46B7iDBr2VJyHUL59QLSWdEAZ5wOoWe8CY/https/i.imgur.com/wSTFkRM.png",
   *     "url": "https://i.imgur.com/wSTFkRM.png",
   *     "width": 0
   *   },
   *   "thumbnail": {
   *       "height": 0,
   *       "proxyURL": "https://images-ext-2.discordapp.net/external/DoAGN014Q46B7iDBr2VJyHUL59QLSWdEAZ5wOoWe8CY/https/i.imgur.com/wSTFkRM.png",
   *       "url": "https://i.imgur.com/wSTFkRM.png",
   *      "width": 0
   *  }
   *  ```
   */
  public getMessageByType(answer: Message, type: messageType) {
    if (type === "embed") {
      const embed = answer.embeds[0];
      if (!embed) {
        return null;
      }
      const tempObject = embed.toJSON() as MinifiedEmbedMessage;
      if (tempObject.image) {
        tempObject.image = pick(tempObject.image, "url");
      }
      if (tempObject.thumbnail) {
        tempObject.thumbnail = pick(tempObject.thumbnail, "url");
      }
      return tempObject;
    } else {
      return answer;
    }
  }
}

const MessageUtils = new MessageUtilsManager();
export default MessageUtils;
