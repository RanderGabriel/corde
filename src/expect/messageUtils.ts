import assert from "assert";
import { Message, MessageEmbed, PartialMessage } from "discord.js";
import {
  MessageIdentifier,
  MessageEmbedLike,
  messageType,
  MinifiedEmbedMessage,
  Primitive,
} from "../types";
import { pick } from "../utils/pick";
import { isPrimitiveValue } from "../utils/isPrimitiveValue";
import { typeOf } from "../utils";

class MessageUtils {
  public messagesMatches(
    returnedMessage: Message | PartialMessage,
    expectation: Primitive | MessageEmbed,
  ) {
    let msg = "";
    if (isPrimitiveValue(expectation)) {
      const formattedMsg = this.getMessageByType(returnedMessage, "text") as Message;
      msg = formattedMsg.content;
      return msg == expectation;
    }

    const jsonMessage = this.getMessageByType(returnedMessage, "embed") as MinifiedEmbedMessage;
    msg = JSON.stringify(jsonMessage);
    let result = true;
    try {
      assert.deepStrictEqual(expectation.toJSON(), jsonMessage);
    } catch (error) {
      result = false;
    }
    return result;
  }

  /**
   * Format Discord responses
   *
   * @param answer Discord response for a message sent
   *
   * @param type Type expected of that message
   *
   * @description Discord adds some attributes that are not present in embed message before it is sent
   *
   *  This is data **before** send to Discord
   *
   *  ```javascript
   *   "image": {
   *       "url": "https://i.imgur.com/wSTFkRM.png"
   *   },
   *   "thumbnail": {
   *       "url": "https://i.imgur.com/wSTFkRM.png"
   *   }
   *  ```
   *
   *  And this is part of embed message **after** get from Discord
   *
   *  ```javascript
   *   "image": {
   *     "height": 0,
   *     "proxyURL": "https://images-ext-2.discordapp.net/external/DoAGN014Q46B7iDBr2VJyHUL59QLSWdEAZ5wOoWe8CY/https/i.imgur.com/wSTFkRM.png",
   *     "url": "https://i.imgur.com/wSTFkRM.png",
   *     "width": 0
   *   },
   *   "thumbnail": {
   *       "height": 0,
   *       "proxyURL": "https://images-ext-2.discordapp.net/external/DoAGN014Q46B7iDBr2VJyHUL59QLSWdEAZ5wOoWe8CY/https/i.imgur.com/wSTFkRM.png",
   *       "url": "https://i.imgur.com/wSTFkRM.png",
   *      "width": 0
   *  }
   *  ```
   */
  public getMessageByType(answer: Message | MessageEmbed | PartialMessage, type: messageType) {
    if (type === "embed") {
      const embed = answer instanceof Message ? answer.embeds[0] : answer;
      if (!embed) {
        return null;
      }

      const tempObject = embed.toJSON() as MinifiedEmbedMessage;
      if (tempObject.image) {
        tempObject.image = pick(tempObject.image, "url");
      }
      if (tempObject.thumbnail) {
        tempObject.thumbnail = pick(tempObject.thumbnail, "url");
      }
      return tempObject;
    } else {
      return answer;
    }
  }

  public createNotFoundMessageForMessageData(msgIdentifier: MessageIdentifier): string | null {
    if (!msgIdentifier) {
      return null;
    }

    if (msgIdentifier.id && msgIdentifier.content) {
      return `Message with id ${msgIdentifier.id} or content '${msgIdentifier.content}' not found.`;
    }

    if (msgIdentifier.id) {
      return `Message with id ${msgIdentifier.id} not found.`;
    }

    if (msgIdentifier.content) {
      return `Message with content '${msgIdentifier.content}' not found.`;
    }

    return null;
  }

  public embedMessageLikeToMessageEmbed(embedLike: MessageEmbedLike) {
    const embed = new MessageEmbed();
    if (typeOf(embedLike) !== "object") {
      return embed;
    }

    if (embedLike.author) {
      if (typeof embedLike.author === "string") {
        embed.setAuthor(embedLike.author);
      } else {
        embed.setAuthor(embedLike.author.name, embedLike.author.iconURL, embedLike.author.url);
      }
    }

    if (embedLike.color) {
      embed.setColor(embedLike.color);
    }

    if (embedLike.description) {
      embed.setDescription(embedLike.description);
    }

    if (embedLike.fields) {
      embed.addFields(...embedLike.fields);
    }

    if (embedLike.files) {
      embed.attachFiles(embedLike.files);
    }

    if (embedLike.footer) {
      if (typeof embedLike.footer === "string") {
        embed.setFooter(embedLike.footer);
      } else {
        embed.setFooter(embedLike.footer.text, embedLike.footer.iconURL);
      }
    }

    if (embedLike.image) {
      if (typeof embedLike.image === "string") {
        embed.setImage(embedLike.image);
      } else {
        embed.setImage(embedLike.image.url);
      }
    }

    if (embedLike.thumbnailUrl) {
      embed.setThumbnail(embedLike.thumbnailUrl);
    }

    if (embedLike.timestamp) {
      embed.setTimestamp(embedLike.timestamp);
    }

    if (embedLike.title) {
      embed.setTitle(embedLike.title);
    }

    if (embedLike.url) {
      embed.setURL(embedLike.url);
    }

    return embed;
  }
}

/**
 * @internal
 */
const messageUtils = new MessageUtils();
export default messageUtils;
