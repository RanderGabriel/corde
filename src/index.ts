/**
 * @file Automatically generated by barrelsby.
 */

 export * from "./cli";
export * from "./erros/index";
