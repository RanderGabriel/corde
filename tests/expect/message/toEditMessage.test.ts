import { Client } from "discord.js";
import MockDiscord from "../../mocks/mockDiscord";
import { createCordeBotWithMockedFunctions } from "../../testHelper";
import { TestReport } from "../../../src/types";
import { ToEditMessage } from "../../../src/expect/matches";
import { buildReportMessage, diff, formatObject } from "../../../src/utils";
import { runtime } from "../../../src/common";
import messageUtils from "../../../src/expect/messageUtils";
import { MockEvents } from "../../mocks/mockEvents";

let mockDiscord = new MockDiscord();

describe("testing toEditMessage", () => {
  afterEach(() => {
    mockDiscord = new MockDiscord();
  });

  it("should return a failed test due to invalid parameter (null)", async () => {
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const message = buildReportMessage(
      `expected: expect value to be a primitive value (string, boolean, number) or an MessageEmbedLike object\n`,
      `received: null`,
    );

    const reportModel: TestReport = {
      pass: false,
      message,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action(null);
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test due to invalid parameter (undefined)", async () => {
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const message = buildReportMessage(
      `expected: expect value to be a primitive value (string, boolean, number) or an MessageEmbedLike object\n`,
      `received: undefined`,
    );

    const reportModel: TestReport = {
      pass: false,
      message,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action(undefined);
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to message was not edited by the bot (no messageIdentifier)", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const message = buildReportMessage(
      `expected: testing bot to edit the last message sent\n`,
      `received: message was not edited`,
    );

    const reportModel: TestReport = {
      pass: false,
      message,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action("pong");
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to message was not edited by the bot (no messageIdentifier as string)", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const message = buildReportMessage(
      `expected: testing bot to edit the message of id 123\n`,
      `received: message was not edited`,
    );

    const reportModel: TestReport = {
      pass: false,
      message,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action("pong", "123");
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to message was not edited by the bot (no messageIdentifier as object with id)", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const message = buildReportMessage(
      `expected: testing bot to edit the message of id 123\n`,
      `received: message was not edited`,
    );

    const reportModel: TestReport = {
      pass: false,
      message,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action("pong", { id: "123" });
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to message was not edited by the bot (no messageIdentifier as object with content)", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const message = buildReportMessage(
      `expected: testing bot to edit the message of content "message test"\n`,
      `received: message was not edited`,
    );

    const reportModel: TestReport = {
      pass: false,
      message,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action("pong", { oldContent: "message test" });
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to timeout but isNot = true", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const reportModel: TestReport = {
      pass: true,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", true);
    const report = await toEditMessage.action("pong");
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned equal message", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange();

    const reportModel: TestReport = {
      pass: true,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action(mockDiscord.message.content);

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned equal messages (string type)", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange();

    const reportModel: TestReport = {
      pass: true,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action(mockDiscord.message.content);

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned equal messages (string number)", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());
    mockDiscord.message.content = "2";

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange();

    const reportModel: TestReport = {
      pass: true,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action(2);

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned equal messages (type embed)", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange(mockDiscord.messageWithEmbed);

    const reportModel: TestReport = {
      pass: true,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action(mockDiscord.messageEmbedLike);

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get failed test due to bot returned equal messages (isNot true)", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange(mockDiscord.messageWithEmbed);

    const reportModel: TestReport = {
      pass: false,
      message: buildReportMessage(
        `expected: message from bot be different from expectation\n`,
        `received: both returned and expectation are equal`,
      ),
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", true);
    const report = await toEditMessage.action(mockDiscord.messageEmbedLike);

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned different messages (isNot true)", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange();

    const reportModel: TestReport = {
      pass: true,
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", true);
    const report = await toEditMessage.action(mockDiscord.messageEmbedLike);

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get fail test due to bot returned different messages (both embed)", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange(mockDiscord.messageWithEmbed);

    const embedReturned = messageUtils.getMessageByType(mockDiscord.messageWithEmbed, "embed");
    const embedLike = {
      author: "Test",
      fields: mockDiscord.messageEmbedLike.fields,
    };

    const embedExpect = messageUtils.embedMessageLikeToMessageEmbed(embedLike);
    const embedExpectedMinified = messageUtils.getMessageByType(embedExpect, "embed");

    const reportModel: TestReport = {
      pass: false,
      message: buildReportMessage(diff(embedReturned, embedExpectedMinified)),
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action(embedLike);

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get fail test due to bot returned different messages (expect primitive and returned embed)", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange(mockDiscord.messageWithEmbed);

    const expectValue = "expect value";
    const embedReturned = messageUtils.getMessageByType(mockDiscord.messageWithEmbed, "embed");

    const reportModel: TestReport = {
      pass: false,
      message: buildReportMessage(
        `expected: '${expectValue}'\n`,
        `received: ${formatObject(embedReturned)}`,
      ),
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action(expectValue);

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get fail test due to bot returned different messages both primitive values", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange();

    const expectValue = "expect value";

    const reportModel: TestReport = {
      pass: false,
      message: buildReportMessage(
        `expected: '${expectValue}'\n`,
        `received: '${mockDiscord.message.content}'`,
      ),
    };

    const toEditMessage = new ToEditMessage(cordeClient, "ping", false);
    const report = await toEditMessage.action(expectValue);

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });
});
