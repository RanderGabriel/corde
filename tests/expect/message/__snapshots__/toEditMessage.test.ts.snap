// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing toEditMessage should fail due to message was not edited by the bot (no messageIdentifier as object with content) 1`] = `
Object {
  "message": "    
    expected: testing bot to edit the message of content \\"message test\\"
    received: message was not edited
",
  "pass": false,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should fail due to message was not edited by the bot (no messageIdentifier as object with id) 1`] = `
Object {
  "message": "    
    expected: testing bot to edit the message of id 123
    received: message was not edited
",
  "pass": false,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should fail due to message was not edited by the bot (no messageIdentifier as string) 1`] = `
Object {
  "message": "    
    expected: testing bot to edit the message of id 123
    received: message was not edited
",
  "pass": false,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should fail due to message was not edited by the bot (no messageIdentifier) 1`] = `
Object {
  "message": "    
    expected: testing bot to edit the last message sent
    received: message was not edited
",
  "pass": false,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should get fail test due to bot returned different messages (both embed) 1`] = `
Object {
  "message": "    
    - expected  - 10
    + received  +  8
    
      Object {
        \\"author\\": Object {
    -     \\"icon_url\\": \\"https://i.imgur.com/wSTFkRM.png\\",
    -     \\"name\\": \\"Some name\\",
    -     \\"url\\": \\"https://discord.js.org\\",
    +     \\"icon_url\\": undefined,
    +     \\"name\\": \\"Test\\",
    +     \\"url\\": undefined,
        },
    -   \\"color\\": 39423,
    -   \\"description\\": \\"Some description here\\",
    +   \\"color\\": null,
    +   \\"description\\": null,
        \\"fields\\": Array [
          Object {
            \\"inline\\": false,
            \\"name\\": \\"Regular field title\\",
            \\"value\\": \\"Some value here\\",
          },
          Object {
            \\"inline\\": false,
            \\"name\\": \\"​\\",
            \\"value\\": \\"​\\",
          },
          Object {
            \\"inline\\": true,
            \\"name\\": \\"Inline field title\\",
            \\"value\\": \\"Some value here\\",
          },
          Object {
            \\"inline\\": true,
            \\"name\\": \\"Inline field title\\",
            \\"value\\": \\"Some value here\\",
          },
        ],
        \\"footer\\": null,
        \\"image\\": null,
    -   \\"thumbnail\\": Object {
    -     \\"url\\": \\"https://i.imgur.com/wSTFkRM.png\\",
    -   },
    +   \\"thumbnail\\": null,
        \\"timestamp\\": null,
    -   \\"title\\": \\"Some title\\",
    +   \\"title\\": null,
        \\"type\\": \\"rich\\",
    -   \\"url\\": \\"https://discord.js.org/\\",
    +   \\"url\\": null,
      }
",
  "pass": false,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should get fail test due to bot returned different messages (expect embed and returned primitive) 1`] = `
Object {
  "message": "    
    expected: Object {
      \\"author\\": Object {
        \\"icon_url\\": \\"https://i.imgur.com/wSTFkRM.png\\",
        \\"name\\": \\"Some name\\",
        \\"url\\": \\"https://discord.js.org\\",
      },
      \\"color\\": 39423,
      \\"description\\": \\"Some description here\\",
      \\"fields\\": Array [
        Object {
          \\"inline\\": false,
          \\"name\\": \\"Regular field title\\",
          \\"value\\": \\"Some value here\\",
        },
        Object {
          \\"inline\\": false,
          \\"name\\": \\"​\\",
          \\"value\\": \\"​\\",
        },
        Object {
          \\"inline\\": true,
          \\"name\\": \\"Inline field title\\",
          \\"value\\": \\"Some value here\\",
        },
        Object {
          \\"inline\\": true,
          \\"name\\": \\"Inline field title\\",
          \\"value\\": \\"Some value here\\",
        },
      ],
      \\"footer\\": null,
      \\"image\\": null,
      \\"thumbnail\\": Object {
        \\"url\\": \\"https://i.imgur.com/wSTFkRM.png\\",
      },
      \\"timestamp\\": null,
      \\"title\\": \\"Some title\\",
      \\"type\\": \\"rich\\",
      \\"url\\": \\"https://discord.js.org/\\",
    }
    received: 'this is the message content'
",
  "pass": false,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should get fail test due to bot returned different messages (expect primitive and returned embed) 1`] = `
Object {
  "message": "    
    expected: 'expect value'
    received: Object {
      \\"author\\": Object {
        \\"icon_url\\": \\"https://i.imgur.com/wSTFkRM.png\\",
        \\"name\\": \\"Some name\\",
        \\"url\\": \\"https://discord.js.org\\",
      },
      \\"color\\": 39423,
      \\"description\\": \\"Some description here\\",
      \\"fields\\": Array [
        Object {
          \\"inline\\": false,
          \\"name\\": \\"Regular field title\\",
          \\"value\\": \\"Some value here\\",
        },
        Object {
          \\"inline\\": false,
          \\"name\\": \\"​\\",
          \\"value\\": \\"​\\",
        },
        Object {
          \\"inline\\": true,
          \\"name\\": \\"Inline field title\\",
          \\"value\\": \\"Some value here\\",
        },
        Object {
          \\"inline\\": true,
          \\"name\\": \\"Inline field title\\",
          \\"value\\": \\"Some value here\\",
        },
      ],
      \\"footer\\": null,
      \\"image\\": null,
      \\"thumbnail\\": Object {
        \\"url\\": \\"https://i.imgur.com/wSTFkRM.png\\",
      },
      \\"timestamp\\": null,
      \\"title\\": \\"Some title\\",
      \\"type\\": \\"rich\\",
      \\"url\\": \\"https://discord.js.org/\\",
    }
",
  "pass": false,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should get fail test due to bot returned different messages both primitive values 1`] = `
Object {
  "message": "    
    expected: 'expect value'
    received: 'this is the message content'
",
  "pass": false,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should get failed test due to bot returned equal messages (isNot true) 1`] = `
Object {
  "message": "    
    expected: message from bot be different from expectation
    received: both returned and expectation are equal
",
  "pass": false,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should get success test due to bot returned different messages (isNot true) 1`] = `
Object {
  "pass": true,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should get success test due to bot returned equal message 1`] = `
Object {
  "pass": true,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should get success test due to bot returned equal messages (string number) 1`] = `
Object {
  "pass": true,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should get success test due to bot returned equal messages (string type) 1`] = `
Object {
  "pass": true,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should get success test due to bot returned equal messages (type embed) 1`] = `
Object {
  "pass": true,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should get success test due to timeout but isNot = true 1`] = `
Object {
  "pass": true,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should return a failed test due to invalid parameter (null) 1`] = `
Object {
  "message": "    
    expected: expect value to be a primitive value (string, boolean, number) or an MessageEmbedLike object
    received: null
",
  "pass": false,
  "testName": "toEditMessage",
}
`;

exports[`testing toEditMessage should return a failed test due to invalid parameter (undefined) 1`] = `
Object {
  "message": "    
    expected: expect value to be a primitive value (string, boolean, number) or an MessageEmbedLike object
    received: undefined
",
  "pass": false,
  "testName": "toEditMessage",
}
`;
